File: Matlab/Epanechnikov/combler_trous.m
function matrice = combler_trous(matr_estim1,matr_estim2,pas_echantillonnage)
% MATRICE = COMBLER_TROUS(MATR_ESTIM1,MATR_ESTIM2,PAS_ECHANTILLONNAGE)
%   comble les "trous" dans la trajectoire. Les trous correspondent à des
%   endroits ou aucune estimation n'a été calculé à cause d'un trop petit
%   nombre de données
%
%   INPUT PARAMETERS:
%   MATR_ESTIM1: Matrice correspondant aux localisations après la
%               suppression des outliers
%   MATR_ESTIM2: Matrice correspondant aux localisations après la deuxième
%               estimation (lissage et réechantillonnage)
%   PAS_ECHANTILLONNAGE: Pas d'echantillonage de la deuxième matrice
%
%   OUTPUT PARAMETERS:
%   MATRICE: Matrice résultante. Dans les "trous" on calcule un troncon de
%           trajectoire en ligne droite
%
%   AUTHOR  :   Julien VANDERSTRAETEN
%   DATE    :   08/2005

% Initialisation des variables du programme
Tps = 1;
Lat = 2;
Lon = 3;
NQ = 4;

% Récupérer la taille de la matrice
[nb_lig_matr_estim1,nb_col_matr_estim1] = size(matr_estim1);
[nb_lig_matr_estim2,nb_col_matr_estim2] = size(matr_estim2);

cpt=1;

% Recopier la premiere loc
matrice(1,:) = matr_estim2(1,:);

% Pour chaque loc. (sauf la première)
for i=2:nb_lig_matr_estim2
    % Si il y a un "trou" dans l'echantillonnage
    if (matr_estim2(i,Tps) - matr_estim2(i-1,Tps)) > pas_echantillonnage
        % Pour chaque estimation manquante dans cette intervalle
        for j=pas_echantillonnage:pas_echantillonnage:(matr_estim2(i,Tps) - matr_estim2(i-1,Tps) - 1)
            tps_esti = j + matr_estim2(i-1,Tps);
            lat1 = matrice(cpt,Lat);
            lon1 = matrice(cpt,Lon);
            tps1 = matrice(cpt,Tps);
            
            % Trouver le point suivant dans la premiere estimation (sans
            % les outliers). L'indice de ce point est donné par k
            for k=1:nb_lig_matr_estim1
                if matr_estim1(k,Tps) > tps_esti
                    if matr_estim1(k,Tps) < matr_estim2(i,Tps)
                        % Si ce point est d'une assez bonne qualité
                        if (matr_estim1(k,NQ) >= 1)
                            lat2 = matr_estim1(k,Lat);
                            lon2 = matr_estim1(k,Lon);
                            tps2 = matr_estim1(k,Tps);
                        % Sinon
                        else
                            lat2 = matr_estim2(i,Lat);
                            lon2 = matr_estim2(i,Lon);
                            tps2 = matr_estim2(i,Tps);
                        end
                    % Sinon
                    else
                        lat2 = matr_estim2(i,Lat);
                        lon2 = matr_estim2(i,Lon);
                        tps2 = matr_estim2(i,Tps);
                    end
                    break
                end
            end

            % Calcul des coeffs de la droite (a_lat et b_lat) pour la latitude
            a_lat = (lat2 - lat1) / (tps2 - tps1); 
            b_lat = (lat1 * tps2 - lat2 * tps1) / (tps2 - tps1); 
            
            % Calcul des coeffs de la droite (a_lon et b_lon) pour la longitude
            a_lon = (lon2 - lon1) / (tps2 - tps1); 
            b_lon = (lon1 * tps2 - lon2 * tps1) / (tps2 - tps1); 

            esti_lat = a_lat * tps_esti + b_lat;
            esti_lon = a_lon * tps_esti + b_lon;
            
            cpt=cpt+1;
            matrice(cpt,1:3) = [(matr_estim2(i-1,Tps)+j) esti_lat esti_lon];
            % Flag pour preciser que ce n'est pas une estimation
            matrice(cpt,4) = 1;
        end
        cpt = cpt+1;
        matrice(cpt,1:3) = matr_estim2(i,1:3);
        % Flag pour preciser que ce n'est pas une estimation
        matrice(cpt,4) = 1;
    % Sinon on recopie simplement l'observation d'origine
    else
        cpt = cpt+1;
        matrice(cpt,1:3) = matr_estim2(i,1:3);
        % Flag pour preciser que c'est une estimation
        matrice(cpt,4) = 0;
    end
end

% Verifier que la courbe ne s'arrete pas avant la fin car on n'a pas pu
% estimer les derniers points
% Si il manque effectivement des points
if (matr_estim1(nb_lig_matr_estim1,Tps) - matr_estim2(nb_lig_matr_estim2,Tps)) > pas_echantillonnage
    for j=pas_echantillonnage:pas_echantillonnage:(matr_estim1(nb_lig_matr_estim1,Tps) - matr_estim2(nb_lig_matr_estim2,Tps))
        tps_esti = j + matr_estim2(nb_lig_matr_estim2,Tps);
        lat1 = matrice(cpt,Lat);
        lon1 = matrice(cpt,Lon);
        tps1 = matrice(cpt,Tps);

        % Trouver le point suivant dans la premiere estimation (sans
        % les outliers). L'indice de ce point est donnée par k
        for k=1:nb_lig_matr_estim1
            if matr_estim1(k,Tps) > tps_esti
                lat2 = matr_estim1(k,Lat);
                lon2 = matr_estim1(k,Lon);
                tps2 = matr_estim1(k,Tps);
                break
            end
        end

        % Calcul des coeffs de la droite (a et b) pour la latitude
        a_lat = (lat2 - lat1) / (tps2 - tps1);
        b_lat = (lat1 * tps2 - lat2 * tps1) / (tps2 - tps1);

        % Calcul des coeffs de la droite (a et b) pour la longitude
        a_lon = (lon2 - lon1) / (tps2 - tps1);
        b_lon = (lon1 * tps2 - lon2 * tps1) / (tps2 - tps1);

        esti_lat = a_lat * tps_esti + b_lat;
        esti_lon = a_lon * tps_esti + b_lon;

        cpt=cpt+1;
        matrice(cpt,1) = tps_esti;
        matrice(cpt,2) = esti_lat;
        matrice(cpt,3) = esti_lon;
        % Flag pour preciser que ce n'est pas une estimation
        matrice(cpt,4) = 1; 
    end
end

File: Python/Utilities/comblerTrous.py
# -*- coding: utf-8 -*-

import os # On importe le module os
import numpy as np

def comblerTrous(estim1,estim2,pasEchantillonnage,recuperation,convertArrayOfTime,convertSecondToDatetime):

    cpt = 0
    temps1 = convertArrayOfTime(recuperation(estim1,'date'))
    temps1 = map(int,temps1)
    temps2 = convertArrayOfTime(recuperation(estim2,'date'))    
    temps2 = map(int,temps2)
    temps1 = map(int,convertArrayOfTime(recuperation(estim1,'date')))
    temps2 = map(int,convertArrayOfTime(recuperation(estim2,'date')))    
    classe = recuperation(estim1,'LC')
    latEstim1 = map(float,recuperation(estim1,'lat'))
    lonEstim1 = map(float,recuperation(estim1,'lon'))
    latEstim2 = map(float,recuperation(estim2,'lat'))
    lonEstim2 = map(float,recuperation(estim2,'lon'))

    matr=np.zeros((10000,3))

    for l in xrange(len(classe)):
        if (classe[l]=='A'):
            classe[l]=-1.0
        
        elif(classe[l]=='B'):
            classe[l]=-2.0
        
        elif(classe[l]=='Z'):
            classe[l]=-3.0
        
        else:
            classe[l]=float(classe[l]) 


    matr[0][0] = temps2[0]
    matr[0][1] = latEstim2[0]
    matr[0][2] = lonEstim2[0]


    for i in xrange(1,len(estim2)):

        if(temps2[i] - temps2[i-1])>pasEchantillonnage:

            for j in xrange(pasEchantillonnage,temps2[i]-temps2[i-1]-1,pasEchantillonnage):

                tpsEsti = j + temps2[i-1]
                lat1 = matr[cpt][1]
                lon1 = matr[cpt][2]
                tps1 = matr[cpt][0]

                for k in xrange(len(estim1)):

                    if temps1[k]>tpsEsti:
                        if temps1[k]>temps2[i]:
                            if classe[k]>=1:
                                lat2 = latEstim1[k]
                                lon2 = lonEstim1[k]
                                tps2 = temps1[k]

                            else:
                                lat2 = latEstim2[i]
                                lon2 = lonEstim2[i]
                                tps2 = temps2[i]
                        else:

                            lat2 = latEstim2[i]
                            lon2 = lonEstim2[i]
                            tps2 = temps2[i]

                        break
                            
                aLat = (lat2 - lat1) / (tps2 - tps1)
                bLat = (lat1 * tps2 - lat2 * tps1) / (tps2 - tps1)

                aLon = (lon2 - lon1) / (tps2 - tps1)
                bLon = (lon1 * tps2 - lon2 * tps1) / (tps2 - tps1)

                estiLat = aLat * tpsEsti + bLat     
                estiLon = aLon * tpsEsti + bLon 

                cpt = cpt+1
                
                matr[cpt][0] = temps2[i-1]+j
                matr[cpt][1] = estiLat
                matr[cpt][2] = estiLon

            cpt = cpt+1
            matr[cpt][0] = temps2[i]
            matr[cpt][1] = latEstim2[i]
            matr[cpt][2] = lonEstim2[i]

        else:

            cpt = cpt+1
            matr[cpt][0] = temps2[i]
            matr[cpt][1] = latEstim2[i]
            matr[cpt][2] = lonEstim2[i]

    if (temps1[len(estim1)-1] - temps2[len(estim2)-1])>pasEchantillonnage:

        for j in xrange(pasEchantillonnage,temps1[len(estim1)-1] - temps2[len(estim2)-1],pasEchantillonnage):

            tpsEsti = j+temps2[len(estim2)-1]
            lat1 = matr[cpt][1]
            lon1 = matr[cpt][2]
            tps1 = matr[cpt][0]

            for k in xrange(len(estim1)):

                if temps1[k]>tpsEsti:

                    lat2 = latEstim1[k]
                    lon2 = lonEstim1[k]
                    tps2 = temps1[k]
                    break
            aLat = (lat2 - lat1) / (tps2 - tps1)
            bLat = (lat1 * tps2 - lat2 * tps1) / (tps2 - tps1)  

            aLon = (lon2 - lon1) / (tps2 - tps1)
            bLon = (lon1 * tps2 - lon2 * tps1) / (tps2 - tps1)

            estiLat = aLat * tpsEsti + bLat
            estiLon = aLon * tpsEsti + bLon

            cpt = cpt+1
            matr[cpt][0]=tpsEsti
            matr[cpt][1]=estiLat
            matr[cpt][2]=estiLon


    
    nouveListe = []

    for i in xrange(cpt):

        dico ={}
        dico["date"]=convertSecondToDatetime(matr[i][0])
        dico["lat"]=matr[i][1]
        dico["lon"]=matr[i][2]
        nouveListe.append(dico)
    
    return nouveListe   
            



            





                                




